### PR TITLE
feat: add thread serach API

### PR DESCRIPTION
**Pull request by Ali-Salman29**
**Description**

Added the thread search API.

Close https://github.com/edly-io/forum/issues/40.

### CODE REVIEW

File: forum/settings/common.py [hunk 1]
```diff
@@ -13,3 +13,7 @@ def plugin_settings(settings: Any) -> None:
     settings.FORUM_PORT = "4567"
     settings.MONGO_HOST = "mongodb"
     settings.MONGO_PORT = 27017
+
+    settings.ELASTICSEARCH_HOST = "elasticsearch:9200"
+    settings.ELASTICSEARCH_INDEX_NAMES = ["comment_threads", "comments"]
+    settings.MAX_DEEP_SEARCH_COMMENT_COUNT = 1000
```
Comment (taimoor-ahmed-1) [inline forum/settings/common.py:19]: there might be an existing ES config in LMS. lets make it more meaningful
```
settings.FORUM_ELASTICSEARCH_HOST
settings.FORUM_ELASTICSEARCH_INDEX_NAMES
....
```

File: forum/views/search.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
+"""
+Search API Views
+"""
+
+from typing import Any, Dict, List, Optional, Tuple
```
Comment (taimoor-ahmed-1) [inline forum/views/search.py:5]: lets use `dict` instead of `Dict`

Comment (taimoor-ahmed-1) [reply forum/views/search.py:5]: consequently we can remove `Dict` import

Comment (Ali-Salman29) [reply forum/views/search.py:5]: Can we do this in a separate PR? I'll create another PR for this.

File: forum/settings/common.py [hunk 3]
```diff
@@ -13,3 +13,14 @@ def plugin_settings(settings: Any) -> None:
     settings.FORUM_PORT = "4567"
     settings.MONGO_HOST = "mongodb"
     settings.MONGO_PORT = 27017
+
+    settings.FORUM_ELASTICSEARCH_INDEX_NAMES = ["comment_threads", "comments"]
+    settings.FORUM_MAX_DEEP_SEARCH_COMMENT_COUNT = 1000
+
+    settings.ELASTIC_SEARCH_CONFIG = [
```
Comment (taimoor-ahmed-1) [inline forum/settings/common.py:20]: is this config already defined in LMS, can we check that. Alternatively, we can add `FORUM` prefix with this too.

Comment (Ali-Salman29) [reply forum/settings/common.py:20]: Yes, this setting is present in the LMS. I've copied this from the LMS.

File: forum/index_manager.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        response: Dict[str, Any], suggestion_types: List[str]
+    ) -> Optional[str]:
+        """
+        Extract suggestions from the Elasticsearch response.
+        """
+        for suggestion_type in suggestion_types:
+            suggestions: List[Dict[str, Any]] = response.get("suggest", {}).get(
+                suggestion_type, []
+            )
+            if suggestions and suggestions[0]["options"]:
```
Comment (Faraz32123) [inline forum/index_manager.py:92]: we can use `get` method to get options and text
```
options = suggestions and suggestions[0].get("options", [])
        if options:
            return options[0].get("text")
```

File: forum/index_manager.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+            must_clause, filter_clause, sort_criteria
+        )
+
+        if response:
+            return list(
+                {
+                    (
+                        hit["_id"]
+                        if CommentThread.index_name in hit["_index"]
+                        else hit["_source"]["comment_thread_id"]
```
Comment (Faraz32123) [inline forum/index_manager.py:192]: same here, it would be good if use dict get method.

File: forum/models/base_model.py [hunk 6]
```diff
@@ -3,14 +3,16 @@
...
+from typing import Any, List, Optional
 
 from bson import ObjectId
 from pymongo.collection import Collection as PymongoCollection
+from pymongo.command_cursor import CommandCursor
 from pymongo.cursor import Cursor
 
 from forum.mongo import Database, get_database
 
+# Define the Collection type for better type checking
```
Comment (Faraz32123) [inline forum/models/base_model.py:15]: we can remove this comment.

File: forum/serializers/contents.py [hunk 7]
```diff
@@ -125,11 +126,11 @@ class UserContentSerializer(serializers.Serializer[dict[str, Any]]):
     course_id = serializers.CharField()
     anonymous = serializers.BooleanField(default=False)
     anonymous_to_peers = serializers.BooleanField(default=False)
-    created_at = serializers.DateTimeField()
-    updated_at = serializers.DateTimeField()
+    created_at = CustomDateTimeField()
```
Comment (Faraz32123) [inline forum/serializers/contents.py:129]: `created_at = serializers.DateTimeField(format="%Y-%m-%dT%H:%M:%SZ")`
While saving the date, we can use datetime.now().strftime("%Y-%m-%dT%H:%M:%SZ")
we will not need to make CustomDateTimeField for this then.

Comment (Ali-Salman29) [reply forum/serializers/contents.py:129]: There are multiple date fields in the different models therefore I've created a separate PR. Moreover, %S doesn't format the date as in the v1 requests. We have to truncate it to 2 decimal places.

**Review by taimoor-ahmed-1 — APPROVED** [review]

**Pull request merged**